README
Spread-Client
==========================

Perl implementation of a Spread Toolkit client.  It allows asynchronous
operations for everything EXCEPT connects.

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

See Makefile.PL

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2008 by Marlon Bailey

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Spread::Client.

0.01  Fri Jul 25 13:33:40 2008
	- original version; created by h2xs 1.23 with options
		-b 5.6.0 -X Spread::Client



lib/Spread/Client/Frame.pm
package Spread::Client::Frame;
use strict;
use warnings;
use Spread::Client::Constant qw(REG_MEMB_MESS
                                REGULAR_MESS
                                TRANSITION_MESS
                                MEMBERSHIP_MESS
                                SPREAD_VERSION
                                CAUSED_BY_LEAVE);

# All lengths are calculated in bytes
use bytes;
use Exporter 'import';

# our EXPORT_OK;
# TODO considering moving this over to EXPORT_OK
# since it might make sense to push this out aa a seperate thing
our @EXPORT= qw(build_connect_message
                build_auth_message
                build_standard_message
                parse_message_header
                parse_message_groups
                parse_message_body
                JOIN_MESS 
                LEAVE_MESS 
                KILL_MESS
                DEFAULT_SEND_MESS
               );

# HEADER INFORMATION INDEXES
sub SERVICE_TYPE () { 0 }
sub SENDER       () { 1 }
sub NUM_MEMBERS  () { 2 }
sub HINT         () { 3 }
sub DATA_LEN     () { 4 }

# version constants
sub MAJOR_VERSION () { 4 }
sub MINOR_VERSION () { 0 }
sub PATCH_VERSION () { 0 }
sub PRIORITY      () { 0 }

# version bitmasks
sub SPREAD3 () { 0x03000000 }
sub SPREAD4 () { 0x04000000 }

# Internal messaging constants
sub JOIN_MESS ()         { 0x00010000 }
sub LEAVE_MESS ()        { 0x00020000 }
sub KILL_MESS ()         { 0x00020000 }
sub DEFAULT_SEND_MESS () { 0x00000002 } # This is the same as reliable mess

# Template for packing header of each message
my $pack_template = 'C4a32C12';

# Message building functions
sub build_connect_message {
    my %args = @_;

    
    # pack string N5Ca
    # TODO use config for versioning here
    return pack( 'C5a*',
                 4,
                 0,
                 0,
                 1,
                 length( $args{private_name} ),
                 $args{private_name},
               );
}

sub build_auth_message {
    my %args = @_;

    # This should basically end up being NULL ord'ed
    my @auth_method = map { ord $_ } (split //, (split /\s/, $args{auth_method})[0]);

    while (scalar @auth_method < 90) {
        push @auth_method, 0;
    }

    # pack string
    return pack( 'C90',
                 @auth_method,
               );
}

sub build_standard_message {
    my %args = @_;

    my @message; 

    # Add Service type to front of message
    _add_to_message(\@message, $args{type} );

    # Add private group name
    push @message, ($args{from_group});

    # Get the length of to group names
    my $group_len = scalar @{$args{to_groups}};
    my $data_len = $args{data} ? length( $args{data} )
                               : 0;

    # Add to groups, group length
    _add_to_message( \@message,  $group_len);
    # Add message hint
    _add_to_message( \@message, 0 );
    # Add data length
    _add_to_message( \@message, $data_len);

    # Add to group names
    push @message, ( @{ $args{to_groups} } );

    # Add the correct number of addition template formats for pack
    my $full_template = $pack_template . ('a32' x $group_len);

    $full_template .= $data_len ? 'a*'
                                : 'a0';

    if( $data_len ) {
        push @message, ( $args{data} );
    }

    return pack( $full_template, @message);
}

# Converts to integer values to a byte array
sub _add_to_message {
    # $_[0] is a reference, there is no need return it
    for my $bitshift (24, 16, 8, 0) {
        push @{$_[0]}, (($_[1] >> $bitshift) & 0xFF);
    }
}

# Parse functions
sub parse_message_header {

    # $_[0] = buffer
    return unpack('IZ32iii', ${$_[0]});
}

sub parse_message_groups {

    # $_[0] = buffer
    # $_[1] = member count

    return [ unpack('(Z32)' . $_[1], ${$_[0]}) ];
}

sub parse_message_body {

    # $_[0] = buffer
    # $_[1] = service_type

    # We do different things based on what type of message
    if( $_[1] & REG_MEMB_MESS ) {

        my ($proc_id, $time, $index, $vs_group_number) = unpack( 'iiii', ${$_[0]});
        # Spread 4 apparently has somethign in an extra 2 integers
        my (@members);

        if((SPREAD_VERSION + 0) & SPREAD4 ) {
            (@members) = unpack( "x[iiiiii](Z32)$vs_group_number", ${$_[0]});
        }
        else { # Spread 3
            (@members) = unpack( "x[iiii](Z32)$vs_group_number", ${$_[0]});
        }

        # return membership listing as message
        return { proc_id => $proc_id, time => $time, index => $index, members => \@members};
    }
    elsif( $_[1] & REGULAR_MESS) {

        return ${$_[0]};
    }
    elsif( $_[1] & TRANSITION_MESS) {
        my ($proc_id, $time, $index) = unpack( 'iii', ${$_[0]});

        return { proc_id => $proc_id, time => $time, index => $index};
    }
    elsif( $_[1] & MEMBERSHIP_MESS and $_[1] & CAUSED_BY_LEAVE) { # self leave message
        # TODO(NOTE) docs say i should get something in the vs_group_stuff but i'm getting squat
        my ($proc_id, $time, $index, $vs_group_number, $member) = unpack( 'iiiiZ32', ${$_[0]});

        # return membership listing as message
        return { proc_id => $proc_id, time => $time, index => $index, members => [ $member ] };
    }

    # If nothing matches or shouldn't match just return nothing
    return;
}

1;


lib/Spread/Client/Connection.pm
package Spread::Client::Connection;
use strict;
use warnings;

use IO::Socket::INET;
use IO::Socket::UNIX;
use Socket;

sub create_new_socket {
    my $self = shift;
    my %args = @_;

    my ($port, $host) = split(/\@/, $args{spread_name} );
    my $sock;

    if( defined $port and defined $host ) { # INET socket

        $sock = IO::Socket::INET->new( Proto    => 'tcp',
                                       Type     => SOCK_STREAM,
                                    )
                    or die "Could not create IP socket '$port\@$host' because: $!";
    }
    elsif( defined $port ) { # UNIX socket

        $sock = IO::Socket::UNIX->new( Peer     => "/tmp/$port",
                                       Type     => SOCK_STREAM,
                                    )
                    or die "Could not create UNIX socket '/tmp/$port' because: $!\n";
    }
    else {
        die "Not enough information to creata a socket to the Spread daemon\n";
    }

    return $sock;
}

sub is_unix {
    my $self = shift;

    return $self->sock->isa('IO::Socket::UNIX');
}

sub session_connected {
    my $self = shift;

    if( $_[0] ) {
        $self->{session_connected} = $_[0];
    }

    $self->{session_connected};
}

1;


lib/Spread/Client/Connection/Sync.pm
package Spread::Client::Connection::Sync;
use strict;
use warnings;
use bytes;

use base qw/Spread::Client::Connection/;

sub new {
    my $class = shift;

    my $self = {};

    $self->{sock} = $class->create_new_socket( @_ );

    return bless $self, $class;
}

sub is_async { return 0 }

sub sock {
    my $self = shift;

    return $self->{sock};
}

sub read {
    my $self = shift;

    my $buffer;
    my $sock = $self->{sock};
    my ($ret, $total, $len) = (0, 0, $_[0]);

    do {
        $ret = sysread( $sock, $buffer, $len, $ret);

        # TODO handle return exceptions like undef
        $total += $ret;
        $len = $_[0] - $total;
    } while( $len > 0 and !$_[1]);

    return \$buffer;
}

sub write {
    my $self = shift;

    my $sock = $self->{sock};

    # TODO handle return exceptions
    return syswrite( $sock, ${$_[0]}, length ${$_[0]} );
}

sub private_group {
    my $self = shift;

    if( $_[0] ) {
        $self->{private_group} = $_[0];
    }

    return $self->{private_group};
}

sub close {
    my $self = shift;
    my $sock = $self->{sock};

    close $sock;
}

1;


lib/Spread/Client/Connection/Async.pm
package Spread::Client::Connection::Async;
use strict;
use warnings;
use bytes;

# TODO do this smarter
$SIG{PIPE} = 'IGNORE';

use base qw/Spread::Client::Connection/;
use Spread::Client::Frame;
use Spread::Client::Constant ':all';

# States
use constant READING_MESS_HEADER => 0;
use constant READING_GROUPS => 1;
use constant READING_MESSAGE => 2;

# Read sizes indexes
use constant MESS_HEADER_SIZE => 0;
use constant MESS_GROUPS_SIZE => 1;
use constant MESSAGE_SIZE => 2;

# Current Message indexes
use constant SERVICE_TYPE => 0;
use constant SENDER => 1;
use constant GROUPS => 2;
use constant MESS_TYPE => 3;
use constant ENDIAN => 4;
use constant MESSAGE => 5;

#
our %init = ( messages           => [],
              read_sizes         => [48, 0, 0],
              private_group      => undef,
              buffer             => undef,
              messages           => undef,
              current_message    => undef,
              state              => undef,
              message_callback   => undef,
              session_connected  => undef,
            );

sub init {
    my $self = shift;

    $self->{$_} = $init{$_} for keys %init;

    $self->reset_state;
}

sub is_async { return 1 }
sub close { die "ABSTRACT\n" } # close socket
sub read { die "ABSTRACT\n" } # read data off socket; Implement in your derived class
sub write { die "ABSTRACT\n" } # write data to socket; Implement in your derived class
sub sock { die "ABSTRACT\n" } # returns raw socket object; Implement in your derived class
sub watch_read { die "ABSTRACT\n" } # toggle for watching reads on this connection; Implement in your derived class

sub reset_state {
    my $self = shift;

    # TODO presize this since we already know how large it will be
    $self->{current_message} = [];
    $self->{buffer} = '';
    $self->{state} = READING_MESS_HEADER;
    $self->{read_sizes}->[$_] = 0 for (MESS_GROUPS_SIZE, MESSAGE_SIZE);
}

sub message_callback {
    my $self = shift;

    $self->{message_callback} = $_[0]
            if $_[0];

    return $self->{message_callback};
}

sub set_header {
    my $self = shift;

    my ($service_type, $sender, $num_members, $hint, $data_len) = parse_message_header( \$self->{buffer} );

    # Set read sizes from header
    $self->{read_sizes}->[MESS_GROUPS_SIZE] = $num_members * 32;
    $self->{read_sizes}->[MESSAGE_SIZE] = $data_len;

    # Fill in what we have into current message
    $self->{current_message}->[SERVICE_TYPE] = $service_type;
    $self->{current_message}->[SENDER] = $sender;
    $self->{current_message}->[ENDIAN] = undef;
    $self->{current_message}->[MESS_TYPE] = undef;

    $self->{buffer} = substr( $self->{buffer},
                              $self->{read_sizes}->[MESS_HEADER_SIZE],
                            );
}

sub set_groups {
    my $self = shift;

    my $num_of_groups = $self->{read_sizes}->[MESS_GROUPS_SIZE] / 32;

    $self->{current_message}->[GROUPS] = parse_message_groups( \$self->{buffer}, $num_of_groups);

    $self->{buffer} = substr( $self->{buffer},
                              $self->{read_sizes}->[MESS_GROUPS_SIZE],
                            );
}

sub set_message {
    my $self = shift;

    $self->{current_message}->[MESSAGE] = parse_message_body( \$self->{buffer},
                                                              $self->{current_message}->[SERVICE_TYPE],
                                                            );
    $self->{buffer} = substr( $self->{buffer},
                              $self->{read_sizes}->[MESSAGE_SIZE],
                            );
}

sub dispatch_message_and_reset_state {
    my $self = shift;
    my @message;

    $message[$_] = $self->{current_message}->[$_] for (0..5);
    
    $self->reset_state();

    if( $self->{message_callback} ) { 
        $self->{message_callback}->( $self, @message);
    }
    else {
        push @{$self->{messages}}, \@message;
    } 
}

sub private_group {
    my $self = shift;

    $self->{private_group} = $_[0]
            if $_[0];

    return $self->{private_group};
}

sub receive {
    my $self = shift;

    if( $self->{state} == READING_MESS_HEADER ) {

        my $left_to_read = $self->{read_sizes}->[MESS_HEADER_SIZE] - length $self->{buffer};

        if( $left_to_read > 0 ) {
            unless( $self->{buffer} .= ${$self->read( $left_to_read, 1 )} ) {
                $self->close();
            }
        }

        if( length $self->{buffer} >= $self->{read_sizes}->[MESS_HEADER_SIZE]) {
            $self->set_header();
            $self->{state} = READING_GROUPS;
        }
    }

    if( $self->{state} == READING_GROUPS ) {

        my $left_to_read = $self->{read_sizes}->[MESS_GROUPS_SIZE] - length $self->{buffer};

        if( $left_to_read > 0 ) {
            unless( $self->{buffer} .= ${$self->read( $left_to_read, 1 )} ) {
                $self->close();
            }
        }

        if( length $self->{buffer} >= $self->{read_sizes}->[MESS_GROUPS_SIZE]) {
            $self->set_groups();
            $self->{state} = READING_MESSAGE;
        }
    }

    if( $self->{state} == READING_MESSAGE ) {

        my $left_to_read = $self->{read_sizes}->[MESSAGE_SIZE] - length $self->{buffer};

        if( $left_to_read > 0 ) {
            unless( $self->{buffer} .= ${$self->read( $left_to_read, 1 )} ) {
                $self->close();
            }
        }

        if( length $self->{buffer} >= $self->{read_sizes}->[MESSAGE_SIZE] ) {
            $self->set_message();
            $self->dispatch_message_and_reset_state();
        }
    }
}

sub get_queued_messages {
    my $self = shift;

    my $messages = $self->{messages};
    $self->{messages} = [];

    return wantarray ? @$messages
                     : $messages;
}

1;


lib/Spread/Client/Connection/Async/Danga.pm
package Spread::Client::Connection::Async::Danga;
use strict;
use warnings;
use base qw/Danga::Socket Spread::Client::Connection::Async/;

use fields keys %Spread::Client::Connection::Async::init;

sub listen_for_messages {
    my Spread::Client::Connection::Async::Danga $self = shift;
    $self->watch_read( $_[0] || 1 );
}

sub new {
    my Spread::Client::Connection::Async::Danga $class = shift;
    my %args = @_;

    my $self = fields::new( $class )
        unless ref $class;


    $self->SUPER::new( $self->create_new_socket( @_ ) );
    $self->init;

    return $self;
}

sub read {
    my $self = shift;

    my $len = $_[0];
    my $read_buffer;

    do {

        my $buffer = $self->SUPER::read( $len );
        $read_buffer .= $$buffer;
        $len = $_[0] - length $read_buffer;
    } while( $len > 0 and !$_[1]);

    return \$read_buffer;
}

sub event_read {
    my Spread::Client::Connection::Async::Danga $self = shift;
    $self->receive();
}

sub event_hup {
    my Spread::Client::Connection::Async::Danga $self = shift;

    $self->close();
}

sub event_err {
    my Spread::Client::Connection::Async::Danga $self = shift;

    $self->close();
}

1;


lib/Spread/Client/Connection/Async/AnyEvent.pm
package Spread::Client::Connection::Async::AnyEvent;
use strict;
use warnings;
use base qw/Spread::Client::Connection::Async/;
use AnyEvent;
use AnyEvent::Handle;

sub new {
    my $class = shift;
    my %args = @_;

    my $socket = $class->create_new_socket( @_ );
    my $handle =
      AnyEvent::Handle->new (
         fh       => $socket,
         autocork => 0,
    );
    
    $handle->stop_read;

    my $self = { sock => $socket, handle => $handle };
    bless $self, $class;
    $self->init;

    return $self;
}

sub sock {
    my $self = shift;
    return $self->{sock};
}

sub write {
    my $self = shift;
    $self->{handle}->push_write( ${$_[0]} );
}

sub listen_for_messages {
    my $self = shift;

    $self->{handle}->start_read;
    $self->{handle}->on_read( sub { $self->receive } );
}

sub read {
    my $self = shift;

    my $read_buffer;

    if( $self->session_connected ) {
        $read_buffer = substr( $self->{handle}->{rbuf}, 0, $_[0]);
        $self->{handle}->{rbuf} = substr( $self->{handle}->{rbuf}, $_[0]);
    }
    else {
        my $len = $_[0];
        my $sock = $self->sock;

        do {
    
            my $buffer;
            my $ret = sysread( $sock, $buffer, $len);
            $read_buffer .= $buffer;
            $len = $_[0] - length $read_buffer;
        } while( $len > 0 and !$_[1]);
    }

    return \$read_buffer;
}

sub close {
    my $self = shift;
    $self->{sock}->close();
}

1;


lib/Spread/Client.pm
package Spread::Client;

use 5.006;
use strict;
use warnings;

use bytes;
use Spread::Client::Constant ':all';
use Spread::Client::Frame;
use List::Util qw(reduce);
use Socket;

our $VERSION = '0.03_02';

sub connect {
    my %args = @_;

    my ($port, $host) = split /@/, $args{spread_name};
    
    die "Bad/missing SPREAD NAME.\n"
        unless defined $port;

    my $connect_class = $args{connect_class} ? "Spread::Client::Connection::$args{connect_class}"
                                             : "Spread::Client::Connection::Sync";

    eval "require $connect_class";

    die $@
        if $@;

    my $conn = $connect_class->new( @_ )
                or die "Could not connect to spread daemon: $!\n";

    my $sock = $conn->sock;

    # In case connection class sets socket to non-blocking
    # we will set this to non-blocking again below
    unless( defined $sock->blocking( 1 ) ) {
         die "could not set blocking on handle: $!\n";
    }

    # CONNECT ROUTINE
    unless( $conn->is_unix ) {

        my $addr = sockaddr_in($port, inet_aton( $host ));
        connect( $sock, $addr)
            or die "Could not connect to host: $host\n";
    }

    # TODO revisit this at a later date
    my $connect_message = build_connect_message( private_name => $args{private_name} || 0);
    
    # Send connect message
    $conn->write( \$connect_message );
    
    # read buffer
    my $buffer;

    # Get auth methods length, to find what auth methods we have available
    $buffer = $conn->read( 1 );
    my $authlen = ord( $$buffer );
    
    die "Bad authentication length: $authlen\n"
        if $authlen == -1 or $authlen >= 128;
    
    # Get auth methods
    my $auth_method = $conn->read( $authlen );
    
    # Send auth message
    my $auth_method_message = build_auth_message( auth_method => $$auth_method );
    $conn->write( \$auth_method_message );
    
    # Pull down accept and versions and grouplen
    $buffer = $conn->read( 5 );
    
    my (@versions, $accept, $full_version);

    # Get accept
    $accept = unpack('c', $$buffer);

    die "Did not get accept: $accept\n"
        unless $accept == 1;
    #

    $versions[0] = substr( $$buffer, 1, 1); # Major
    $versions[1] = substr( $$buffer, 2, 1); # Minor
    $versions[2] = substr( $$buffer, 3, 1); # Patch

    {
        no warnings 'once';
        $full_version = unpack('c', reduce { $a | $b } (@versions));
    }

    die "Full version is a problem: $full_version\n"
        if $full_version == -1;

    # Get private group info
    # Get group length
    my $grouplen = unpack('c', substr($$buffer, 4, 1));
    
    die "We had an issue with the group length\n"
        if $grouplen == -1;

    # Get actual name private group name
    $buffer = $conn->read( $grouplen );
    my $private_group = unpack("a*", $$buffer);
    $conn->private_group( $private_group );

    if( $conn->is_async ) {
        $sock->blocking( 0 );
        $conn->listen_for_messages;
        $conn->session_connected( 1 );
    }

    return $conn;
}

sub receive {
    my %args = @_;

    if( $args{conn}->is_async ) {

        $args{conn}->receive;

        unless( $args{conn}->message_callback ) {
            return $args{conn}->get_queued_messages;
        }
    }
    else {
        my $buffer;
    
        # pull down message header
        $buffer = $args{conn}->read( 48 );
        
        # define variables we need
        my ($mess_type, $endian);
    
        # unpack header data
        my ($service_type, $sender, $num_members, $hint, $data_len) = parse_message_header( $buffer );
    
        # Get group data
        my $pull_bytes = $num_members * 32;
        $buffer = $args{conn}->read( $pull_bytes );

        my $groups = parse_message_groups( $buffer, $num_members); 
    
        # pull down message body
        $buffer = $args{conn}->read( $data_len );
    
        my $message = parse_message_body( $buffer, $service_type);
    
        return ($service_type, $sender, $groups, $mess_type, $endian, $message);
    }
}

sub multicast {
    my %args = @_;

    my $private_group = $args{conn}->private_group;

    my $multi_message = build_standard_message( type       => $args{type} || DEFAULT_SEND_MESS,
                                                from_group => $private_group,
                                                to_groups  => $args{groups},
                                                data       => $args{message},
                                             );

    my $ret = $args{conn}->write( \$multi_message );

    return $ret;
}

sub join {
    my %args = @_;

    my $private_group = $args{conn}->private_group;

    my $join_message = build_standard_message( type       => JOIN_MESS,
                                               from_group => $private_group,
                                               to_groups  => $args{groups},
                                            );

    my $ret = $args{conn}->write( \$join_message );

    return $ret;
}

sub leave {
    my %args = @_;

    my $private_group = $args{conn}->private_group;

    my $leave_message = build_standard_message( type       => LEAVE_MESS,
                                                from_group => $private_group,
                                                to_groups  => $args{groups},
                                             );

    my $ret = $args{conn}->write( \$leave_message );

    return $ret;
}

sub disconnect {
    my %args = @_;

    my $private_group = $args{conn}->private_group;

    my $disconnect_message = build_standard_message( type       => KILL_MESS,
                                                     from_group => $private_group,
                                                     to_groups  => [ $private_group ],
                                                  );
    # write close message to daemon
    $args{conn}->write( \$disconnect_message );

    # close socket
    $args{conn}->close();
}

1;
__END__

=head1 NAME

Spread::Client - Spread client that allows synchronous OR asynchronous multicast/receive/join/leave/disconnect to spread daemons

=head1 SYNOPSIS

  # ASYNCHRONOUS AnyEvent BEHAVIOR(with POE)
  use strict;
  use warnings;
  use POE;
  use Spread::Client;
  use Spread::Client::Constant ':all'; 
  use Danga::Socket;
  use Data::Dumper;
  
  my $spread_name = '4803@localhost';
  my $private_name = 'mrperla';
  
  
  # set message count before exit
  my $message_max = 10;
  my $message_count = 0;
  
  # Connect using Danga socket connection class for 
  # running with Danga::Socket Event loop
  my $conn = Spread::Client::connect(
        spread_name   => $spread_name,
        private_name  => $private_name,
        connect_class => 'Async::AnyEvent',
     );
  
  # set callback to handle message receipts
  $conn->message_callback( \&handle_message );
  
  Spread::Client::join( conn   => $conn,
                        groups => ['channel-100'],
                      );
  
  # If you decide not to give this message a 'type'
  # then it will default to sending a RELIABLE MESSAGE
  Spread::Client::multicast( conn    => $conn,
                             groups  => ['channel-100'],
                             message => 'this is a message',
                           );

  sub handle_message {
      my ($conn, @message) = @_;
  
      print Dumper( \@message );
  
      $message_count++;
  
      if( $message_count >= $message_max ) {
  
          Spread::Client::leave( conn   => $conn,
                                 groups => ['channel-100'],
                              );
  
          Spread::Client::disconnect( conn => $conn );
          POE::Kernel->stop();
      }
  }

  # Don't forget to start your event loop  
  POE::Kernel->run();

  # ASYNCHRONOUS DANGA BEHAVIOR
  use strict;
  use warnings;
  use Spread::Client;
  use Spread::Client::Constant ':all';
  use Danga::Socket;
  use Data::Dumper;
  
  my $spread_name = '4803@localhost';
  my $private_name = 'mrperla';
  
  
  # set message count before exit
  my $message_max = 10;
  my $message_count = 0;
  
  # Connect using Danga socket connection class for 
  # running with Danga::Socket Event loop
  my $conn = Spread::Client::connect(
        spread_name   => $spread_name,
        private_name  => $private_name,
        connect_class => 'Async::Danga',
     );
  
  # set callback to handle message receipts
  $conn->message_callback( \&handle_message );
  
  Spread::Client::join( conn   => $conn,
                        groups => ['channel-100'],
                      );
  
  # If you decide not to give this message a 'type'
  # then it will default to sending a RELIABLE MESSAGE
  Spread::Client::multicast( conn    => $conn,
                             groups  => ['channel-100'],
                             message => 'this is a message',
                           );

  sub handle_message {
      my ($conn, @message) = @_;
  
      print Dumper( \@message );
  
      $message_count++;
  
      if( $message_count >= $message_max ) {
  
          Spread::Client::leave( conn   => $conn,
                                 groups => ['channel-100'],
                              );
  
          Spread::Client::disconnect( conn => $conn );
          Danga::Socket->SetPostLoopCallback( sub { 0 } );
      }
  }
  # Don't forget to start your event loop  
  Danga::Socket->EventLoop();

  # SYNCHRONOUS BEHAVIOR
  use strict;
  use warnings;
  use Spread::Client;
  use Spread::Client::Constant ':all';
  use Data::Dumper;
  
  my $spread_name = '4803@localhost';
  my $private_name = 'mrperls';
  
  # Connect using Sync class
  # if connect_class is left out, defaults to 'Sync'
  my $conn = Spread::Client::connect(
        spread_name   => $spread_name,
        private_name  => $private_name,
        connect_class => 'Sync',
     );
  
  # If we need the private group
  my $private_group = $conn->private_group;
  
  Spread::Client::join( conn   => $conn,
                        groups => ['channel-100'],
                      );
  
  Spread::Client::multicast( conn    => $conn,
                             type => SAFE_MESS,
                             groups  => ['channel-100'],
                             message => 'this is a message',
                           );
  
  for (1..2) {
  
      my ($service_type, $sender, $groups, $mess_type, $endian, $message) =
      Spread::Client::receive( conn  => $conn );

      # Should Dump a join message, and our multi-cast since SELF_DISCARD isn't on
      warn Dumper( $service_type, $sender, $groups, $mess_type, $endian, $message);
  }
  
  # leave group
  Spread::Client::leave( conn   => $conn,
                         groups => ['channel-100'],
                       );

  # disconnect
  Spread::Client::disconnect( conn => $conn );

=head1 DESCRIPTION

A Spread Toolkit client implemented in perl, that allows both synchronous and asynchronous functionality.
Reading the Spread User's Guide is strongly recommended before using the Spread Toolkit in general.

=head1 CAVEATS

Right now connect is always synchronous, this may change later but unless someone really wants it
i don't see it occuring anytime soon.  Asynchronous is supported for Danga::Socket and AnyEvent which should
cover a large number of the event loops out there, include POE.  If you would like to use AnyEvent or Danga, please have them installed, it's not a prerequisite of the module as connection classes are loaded at runtime and the synchronous connection client does not depend on either Danga::Socket or AnyEvent.

=head1 EXPORT

None by default.


=head1 SEE ALSO

Constants module L<Spread::Client::Constant>.

Without these helpful sources, this module would not have been possible:

Rough spread API, documented by a nice developer. L<http://www.roughtrade.net/spread/spread-client-proto.txt>

Pure Python Spread Client L<http://code.google.com/p/py-spread/>

Spread source including their C and JAVA client L<http://www.spread.org/>

=head1 AUTHOR

Marlon Bailey, E<lt>mbailey@span.org<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by Marlon Bailey

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut
